src/v/storage/segment_appender.h

```diff
@@ -162,21 +205,76 @@ class segment_appender {
     // still heavy weight operations compared to regular flush()
     ss::future<> hard_flush();
 
+    enum write_state { QUEUED = 1, DISPATCHED, DONE };
```

Comment:
Before we only had 2 states: "done" and "not done".

They key here is that writes in QUEUED state haven't been sent anywhere yet so we can still modify them.